Add a warning when casting from capability to pointer (and make C++ implicit capability conversions work like they do in C)

arichardson opened the pull request:
Only (u)intptr_t and vaddr_t (aka __memaddr_t) are acceptable cast targets

This is quite noisy when compiling cheribsd hybrid mode code so I'm not sure I should add this yet.

I also have a change for cheribsd that tags vaddr_t as __attribute__((memory_address)) and fixes a few cast warnings. There are a lot of cases where we are casting capabilities to the corresponging pointer type in the hybrid ABI and I'm not sure how best to silence those? Possibly with a `__builtin_cheri_cap_to_pointer()` builtin?

davidchisnall commented on lib/Sema/Sema.cpp:
```diff
@@ -177,7 +177,7 @@ void Sema::Initialize() {
       PushOnScopeChains(Context.getUIntCapDecl(), TUScope);
   }
-  
+
```
Please remove whitespace-only changes like this one.

brooksdavis reviewed (commented):
I'm not in a position to review the LLVM change, bit I like the idea.  I suspect this will catch most of the `ptrdiff_t` misuse (virtually all `ptrdiff_t` use) and similar things.  Do make sure to test CheriBSD without `NO_WERROR` before commit.

RichardsonAlex commented:
What do you suggest we do about the ``foo* __capability` -> `foo *` casts? Silence the warning or add some explicit casting?

A lot of them are in cheriabi_sysstubs.h, some in libc_c where we do
`void*__capability x = (void*__capability)calloc(...)` and then later `free((void*)x)`

kgudka commented:
I would vote for explicit casting so we know all the places where such a conversion is occurring.

brooksdavis commented:
Hmm.  In the case of `lib/libcheri_syscalls/cheri_syscalls.c` we can change which casts we make, but it would be nice to not have do have an `#ifdef` between hybrid and pure modes.  I suppose that since this is boundary code, it would be ok at have the moral equivalent of `__DECONST()` macros.

davidchisnall commented:
We've so far had a policy of requiring all casts between `X*` and `__capability X*` to be explicit.  Casts from the `__capability` version to the non-`__capability` version are not valid in the general case and so should be checked by a human.  Casts in the other direction are valid, but should come with explicit bounds truncation and so should also be checked by a human.

The same should apply for all casts from capability to and from `__intcap_t`, though casts between `__intcap_t` and integer types are probably fine as implicit casts.

brooksdavis commented:
The things in `cheri_syscalls.c` are of the form:

```C
int
foo_stub(void __capability *arg)
{

    return (foo((void *)arg);
}
```

The `(void *)arg` cast is intended to be a NOP in pure-cability mode and cast away the capability in hybrid mode.  I don't much care how we spell it, but being able to express that in C rather than CPP would be nice.

davidchisnall commented:
(For reference, writing up some of the face-to-face discussion).

We should add explicit cast kinds for these modelled on Objective-C bridging casts.  The following should not cause warnings:

```C
void * __capability cap = ...;
void *x = (__cheri_cast void*)cap;
cap = (__cheri_cast void * __capability)x;
```

This code can then be made compatible with non-CHERI systems with a `#define __cheri_cast` in a header.

We also pondered adding a `__cheri_bounded_cast` for non-capability to capability casts, which would set the bounds of the new pointer to the size of the pointee.  Arrays would need an explicit bounds setting operation (though we might want to accept `(__cheri_bounded_cast struct foo[42])` and `(__cheri_bounded_cast struct foo[x])`to cast to a pointer with a bounded length.

brooksdavis commented:
`__cheri_cast` sounds good to me.

arichardson committed:
Add (__cheri_cast foo*) for explicit capability <-> pointer operations

__cheri_cast can be used to explicitly convert between pointers and
capabilities in the hybrid ABI. It is a noop in purecap mode but is
allowed so that the same code can compile with both hybrid and purecap ABI.

Currently only checking of the source and destination type is implemented.

arichardson committed:
Add a __cheri_cast FixItHint

arichardson committed:
Implement __cheri_cast

For now it just creates a CK_PointerToMemoryCapability or
CK_MemoryCapabilityToPointer CStyleCastExpr. In the future we might want
to add a new __cheri_bounded_cast that also sets the bounds on the
resulting capability

arichardson committed:
Remove debug spam

arichardson committed:
Add a test that __cheri_cast can't remove const

arichardson committed:
Make __capability after the name an error

In principle it should be fine but if we don't it will crash the compiler.

Fixes #134

arichardson committed:
Check the AST in cheri-cap-to-int-cast.c

arichardson committed:
Fix __cheri_cast when used in C++

In C++ mode ASTContext::typesAreCompatible() checks whether the canonical
types are identical. This is not true for capability and non-capability
versions of the same pointer because the end up being a different instance
of clang::PointerType.

arichardson committed:
Add a warning when casting from capability to pointer

Only (u)intptr_t and vaddr_t (aka __memaddr_t) are acceptable cast targets

arichardson committed:
Make the capability to int cast warning work as expected in C++

As long as the source type is a pointer. For some reason it doesn't work
yet with references.

arichardson committed:
Check for capability -> integer narrowing in brace-initilization

This should fix most cases where C++ doesn't perform casting checks
currently done in C

arichardson committed:
Add a test for cap->pointer narrowing in C brace initialization

Narrowing conversions are not normally errors in C but for capabilities I
think it makes sense to be explicit and it is easy to fix with a cast

arichardson committed:
Extend previous test with arrays of structs

arichardson committed:
Make implicit conversions between pointers and capabilities errors in C++

C++ goes through a different code path so the errors we get in C were not
happening when compiling as C++

arichardson committed:
Make implicit conversions from capability to pointer an error

arichardson committed:
Don't warn about capability changes in casts to void

E.g. `(void)(__cheri_cast int* __capability)foo;` would previously warn

arichardson committed:
Update two tests after the latest changes

arichardson committed:
Move the capability conversion check a bit later

Otherwise we replace some more useful errors with the generic bad
initialization error

arichardson committed:
Don't diagnose cap <-> pointer conversions in purecap array to pointer decay

In that case the original type will not be capability qualified but the
array to pointer decay will result in a char[N] -> char * __capability
conversion.
Fixes SemaCXX/cheri-template-array-size.cpp

arichardson committed:
Fix null pointer dereference in some test cases

arichardson committed:
Allow casting of nullptr to integers also in purecap mode

arichardson committed:
Fix the cheri-pointer-cast.cpp test case

arichardson committed:
Add commented out test for #147

arichardson committed:
Adjust to function renaming

arichardson committed:
Make cheri-cap-to-int-cast.c pass again after latest renames

arichardson committed:
Add a better error message when assigning better capability and pointer

This also means that the error message is the same between C and C++,
making it easier to test that we reject these assignments in both cases

arichardson committed:
Update cheri-cap-to-int-cast.cpp after latest changes to diagnostics

RichardsonAlex commented:
This now also contains commits that disallow implicit conversions from capabilities to pointer/pointers to capabilities in C++ because splitting that into a separate pull request would require some nontrivial history rewriting

arichardson committed:
Allow implicit conversion from string literals to capabilities

arichardson committed:
Fix the cast kind for string literal -> capability conversion

arichardson committed:
Extend last test to also check that we can call functions with string literals

arichardson committed:
Don't warn about capability->pointer conversions constructor calls

A C++ functional cast could also be a constructor call so don't warn in
that case

arichardson committed:
Remove an incorrect narrowing conversion check for array elements

Fixes cases where an initializer list calls an implicit constructor that
takes a capability pointer. This fixes compilation of libc++ memory.cpp

arichardson committed:
Extend cheri-brace-init.c for purecap ABI

arichardson committed:
Make the capability to int warning a -pedantic warning

arichardson committed:
Don't reuse the DeprecatedStringLiteralToCharPtr field in Overload.h

This fixes some incorrect warnings about using __cheri_cast instead of
printing the deprecated conversion warning.
To find all the calls to setStandard() I added an new overload that uses
memset() to clear all fields before using it.

arichardson committed:
Don't warn about capability to integer conversion for dependent types

Dependent types are never capability types so this warning will always
trigger in template definitions

arichardson committed:
[UPSTREAM] Use ImplicitConversionSequence::setAsIdentityConvesion(). NFC

arichardson committed:
Add __has_feature(__cheri_cast)

So that old code can #define it to nothing for compatibility

arichardson committed:
Turn capability-to-integer-cast warning off by default

It is extremely noisy and only sometimes finds real issues

arichardson committed:
Make capability to plain pointer cast a separate warning

This one is much less likely to be a false positive compared to casting
to an integer. I think all these cases are actually errors so I made it
an error by default

arichardson committed:
Read CLANG_FORCE_COLOR_DIAGNOSTICS env var to determine colour diagnostics default

This allows forcing colour diagnostics to on for projects where adding
extra compiler flags to the build system is not possible.

arichardson committed:
Don't strip __capability from reference types in Desugar()

Otherwise diagnostics print `'CapRef' (aka 'int &')` instead of
`'CapRef' (aka 'int & __capability')`

arichardson committed:
Make implicit conversions between int&__capability and int& an error

Also slightly changed the error message to work for references

arichardson committed:
Also diagnose int& -> int& __capability conversions in casts

davidchisnall commented on include/clang/Basic/Attr.td:
```diff
@@ -1621,6 +1621,10 @@ def CHERICapability : TypeAttr {
   let Documentation = [Undocumented];
 }
+def MemoryAddress : TypeAttr {
+  let Spellings = [GNU<"memory_address">];
```
Please can we have a C++ attribute for this as well?

davidchisnall commented on include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -2453,6 +2453,9 @@ def err_attribute_pointers_only : Error<warn_attribute_pointers_only.Text>;
   "%0 attribute argument may only refer to a function parameter of integer "
   "type">;
+def err_attribute_address_integers_only : Error<
+  "%0 attribute only applies to integer types that can store memory "
```
Perhaps 'can store addresses' would be better?

davidchisnall commented on include/clang/Basic/TokenKinds.def:
```diff
@@ -563,6 +563,11 @@ KEYWORD(__bridge_transfer            , KEYARC)
 
+
+// CHERI casts to and from capabilities
+KEYWORD(__cheri_cast                 , KEYALL)
```
It would be nice to have an `__offset_cast` and `__address_cast` for capability-to-integer casts as well.

Should this be a keyword for all language dialects, or only for CHERI?  Ideally, it would be a normal identifier in other language dialects, though that's not essential because it's reserved for the implementation.

davidchisnall commented on include/clang/Sema/Overload.h:
```diff
@@ -193,6 +193,17 @@ namespace clang {
     unsigned ObjCLifetimeConversionBinding : 1;
+
+    /// \brief e.g. conversions from pointer -> capability without an explicit __cheri_cast
+    unsigned IncompatibleCHERIConversion : 1;  // XXXAR: would be nice if we had a ctor to initialize this
```
Does `= 0` not do the right thing WRT constructors?  Also, is there a reason that this is not a `bool`?

RichardsonAlex replied:
`= 0` doesn't work because this type can't have a non-trivial constructor due to the way it is used. Also I'm not sure it works for bitfields before C++17

davidchisnall commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -951,11 +951,18 @@ static bool parseShowColorsArgs(const ArgList &Args, bool DefaultColor) {
+            .Cases("no", "0", Colors_Off)
+            .Cases("yes", "always", "1", Colors_On)
+            .Default(Colors_Auto);
+  }
```
This would be good to upstream - it looks generally useful, especially for ninja and friends.

davidchisnall commented on lib/Sema/SemaCast.cpp:
```diff
@@ -1941,7 +2018,6 @@ static TryCastResult TryReinterpretCast(Sema &Self, ExprResult &SrcExpr,
                                         unsigned &msg,
                                         CastKind &Kind) {
   bool IsLValueCast = false;
-  
```
This and the next pure-whitespace change would be good to remove, if possible.

davidchisnall commented on lib/Sema/SemaCast.cpp:
```diff
@@ -2717,8 +2794,77 @@ ExprResult Sema::BuildCXXFunctionalCastExpr(TypeSourceInfo *CastTypeInfo,
+    return ExprError();
+  }
+  bool TypesCompatible = false;
+  if (getLangOpts().CPlusPlus) {
```
LLVM style doesn't want these braces (or the ones around the else).

davidchisnall commented on lib/Sema/SemaCast.cpp:
```diff
@@ -2717,8 +2794,77 @@ ExprResult Sema::BuildCXXFunctionalCastExpr(TypeSourceInfo *CastTypeInfo,
+    Kind = CK_CHERICapabilityToPointer;
+    assert(!Context.getTargetInfo().areAllPointersCapabilities() &&
+           "__cheri_cast to pointer should not be possible in purecap mode");
+  } else if (DestIsCap && !SrcIsCap) {
```
LLVM style doesn't want these braces either.

davidchisnall commented on lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -6612,6 +6612,10 @@ static void ProcessDeclAttribute(Sema &S, Scope *scope, Decl *D,
+//     D->dump(llvm::errs());
+//     assert(false && "AT_MEMORY_ADDRESS in declattr");
+//     break;
   }
```
Please either remove or uncomment (and document) this code.

davidchisnall commented on lib/Sema/SemaExpr.cpp:
```diff
@@ -7698,21 +7698,25 @@ Sema::CheckAssignmentConstraints(QualType LHSType, ExprResult &RHS,
+      if (AddrSpaceL != AddrSpaceR) {
         Kind = CK_AddressSpaceConversion;
-      else if (LHSPointer->isFunctionPointerType() && RHSPointer->isFunctionPointerType()) {
+      } else if (LHSPointer->isFunctionPointerType() && RHSPointer->isFunctionPointerType()) {
```
Change violates LLVM style.

arichardson committed:
Add a use __builtin_assume_aligned note to the cast alignment error

arichardson committed:
Remove use __cheri_cast note and add it to the error message instead

davidchisnall commented on lib/Sema/SemaOverload.cpp:
```diff
@@ -1248,7 +1253,7 @@ TryUserDefinedConversion(Sema &S, Expr *From, QualType ToType,
         // gets ranked with standard conversion sequences.
         DeclAccessPair Found = ICS.UserDefined.FoundConversionFunction;
-        ICS.setStandard();
+        ICS.setStandard(ImplicitConversionSequence::MemsetToZero);
```
I'm not sure any of these are needed if you simply write `= 0` to initialise your new field.

davidchisnall commented on lib/Sema/SemaOverload.cpp:
```diff
@@ -1563,6 +1568,7 @@ static bool IsStandardConversion(Sema &S, Expr* From, QualType ToType,
   SCS.IncompatibleObjC = false;
   SCS.setFromType(FromType);
   SCS.CopyConstructor = nullptr;
+  // assert(!SCS.IncompatibleCHERIConversion);
```
Please uncomment or remove.

davidchisnall reviewed (approved):
Some small issues, but I'm happy for them to be addressed after merging.

arichardson committed:
Fix checks for casts betwen non-cap and cap references

arichardson committed:
Fix taking references to __intcap_t in the hybrid ABI

The reference intialization changes made this an error because I was using
isCHERICapabilityType(S.getASTContext()) which is also true for __intcap_t

arichardson committed:
Fix diagnosing cast between cap and non-cap references in purecap ABI

arichardson committed:
Add the required -Wno-error to fix codegen test

arichardson committed:
Fix the final CodeGen test after __cheri_cast changes

arichardson committed:
Add a C++11 spelling of the memory_address attribute

arichardson committed:
Slightly adjust the error messsage for err_attribute_address_integers_only

arichardson committed:
Style fixes and remove commented out debug code. NFC

arichardson committed:
Fix null pointer dereference
